pkg/db/subscription.go

```diff
@@ -0,0 +1,101 @@
+package db
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type PollableDBQuery[ValueType any] func(ctx context.Context, lastSeenID int64, numRows int32) (results []ValueType, lastID int64, err error)
+
+// Poll whenever notified, or at an interval if not notified
+type PollingOptions struct {
+	Interval time.Duration
+	Notifier <-chan bool
+	NumRows  int32
+}
+
+type DBSubscription[ValueType any] struct {
+	ctx        context.Context
+	log        *zap.Logger
+	lastSeenID int64
+	options    PollingOptions
+	query      PollableDBQuery[ValueType]
+	updates    chan<- []ValueType
+}
+
+func NewDBSubscription[ValueType any](
+	ctx context.Context,
+	log *zap.Logger,
+	query PollableDBQuery[ValueType],
+	lastSeenID int64,
+	options PollingOptions,
+) *DBSubscription[ValueType] {
+	return &DBSubscription[ValueType]{
+		ctx:        ctx,
+		log:        log,
+		lastSeenID: lastSeenID,
+		options:    options,
+		query:      query,
+		updates:    nil,
+	}
+}
+
+func (s *DBSubscription[ValueType]) Start() (<-chan []ValueType, error) {
+	if s.updates != nil {
+		return nil, fmt.Errorf("Already started")
+	}
+	if s.options.NumRows <= 0 || s.log == nil {
+		return nil, fmt.Errorf("Required params not provided")
+	}
+	updates := make(chan []ValueType)
+	s.updates = updates
+
+	go func() {
+		s.poll()
+
+		timer := time.NewTimer(s.options.Interval)
+		for {
+			timer.Reset(s.options.Interval)
+			select {
+			case <-s.ctx.Done():
+				s.log.Info("Context done; stopping subscription")
+				close(s.updates)
+				return
+			case <-s.options.Notifier:
```

Comment:
I think you can reset the timer here, so that if the timer is a 2s interval and you get notified after 1.8s, it won't check again in 0.2s

Reply:
See other comment!